runtime/signals.c

```diff
@@ -94,16 +86,13 @@ if (!check_for_pending_signals())
     if( specific_signal_pending > 0 ) {
       if( !atomic_compare_exchange_strong(
             &caml_pending_signals[i],
-             &specific_signal_pending, specific_signal_pending - 1) ) {
+             &specific_signal_pending, 0) ) {
         /* We failed our CAS because another thread beat us to processing
            this signal. Try again to see if there are more of this signal
            to process. */
         goto again;
       }
```

Comment:
You might be able to simplify the whole `again: ... goto again` by using a single atomic_exchange, to read the old value and `caml_pending_signals[i]` and set it to 0 atomically.  On the other hand, the current code might be more efficient if `caml_pending_signals[i]` is zero most of the time, which it should be in practice.  Your choice.

Reply:
Exchange does simplify the code quite nicely but having a ponder I think CAS is better here. I worry about multiple sockets playing cache line ping-pong as a few hundred threads come out of a stop-the-world pause and all race to deliver a waiting signal.